main.go

```diff
@@ -450,15 +452,37 @@ func (s *site) nextProxy() *Backend {
 	if len(backends) == 0 {
 		return nil
 	}
-
-	idx := rand.Intn(len(backends))
+	min := math.MaxInt32
+	earliest := time.Now().Add(time.Second)
+	idx := 0
+	for i, backend := range backends {
+		backend.Stats.RLock()
+		if backend.Stats.CurrentCalls < min {
+			min = backend.Stats.CurrentCalls
+			if backend.Stats.LastFinished.Before(earliest) {
+				earliest = backend.Stats.LastFinished
+				idx = i
+			}
+		}
+		backend.Stats.RUnlock()
+	}
 	// random backend from a list of available backends.
-	return backends[idx]
+	backend := backends[idx]
+	backend.Stats.Lock()
+	backend.Stats.CurrentCalls++
+	backend.Stats.Unlock()
+	return backend
 }
 
 // ServeHTTP - LoadBalancer implements http.Handler
 func (s *site) ServeHTTP(w http.ResponseWriter, r *http.Request) {
 	backend := s.nextProxy()
+	defer func() {
```

Comment:
`nil` backend must be handled. The deferred function is before that is checked.

```
backend, done := s.nextProxy()
if backend == nil || !backend.Online() {...}
defer done()
```

I don't really know why we return offline backends in the first place.

Reply:
yeah. now is fine? @klauspost

Reply:
Looks great!